contracts/badger-sett/strategies/unitprotocol/StrategyUnitProtocolMeta.sol

```diff
@@ -0,0 +1,248 @@
+        _safeApproveHelper(collateral, unitVault, _collateralAmt);
+        IUnitCDPManager(cdpMgr01).join(collateral, _collateralAmt, _usdpAmt);
+    }
+
+    function repayAndRedeemCollateral(uint256 _collateralAmt, uint256 _usdpAmt) internal {
+        _safeApproveHelper(debtToken, unitVault, _usdpAmt);
+        IUnitCDPManager(cdpMgr01).exit(collateral, _collateralAmt, _usdpAmt);
+    }
+
+    // **** State Mutation functions ****
+
+    function keepMinRatio() external {
+        _onlyCDPInUse();
+        _onlyAuthorizedActorsOrController();
+
+        uint256 requiredPaidback = requiredPaidDebt(0);
+        if (requiredPaidback > 0) {
+            _withdrawUSDP(requiredPaidback);
+
+            uint256 _actualPaidDebt = IERC20Upgradeable(debtToken).balanceOf(address(this));
+            uint256 _fee = getDebtBalance().sub(getDebtWithoutFee());
+
+            require(_actualPaidDebt > _fee, "!notEnoughForFee");
+            _actualPaidDebt = _actualPaidDebt.sub(_fee); // unit protocol will charge fee first
+            _actualPaidDebt = _capMaxDebtPaid(_actualPaidDebt);
+
+            require(IERC20Upgradeable(debtToken).balanceOf(address(this)) >= _actualPaidDebt.add(_fee), "!notEnoughRepayment");
+            repayAndRedeemCollateral(0, _actualPaidDebt);
+        }
+    }
+
+    /// @dev Internal deposit logic to be implemented by Stratgies
+    function _deposit(uint256 _want) internal override {
+        if (_want > 0) {
+            uint256 _newDebt = calculateDebtFor(_want.add(getCollateralBalance()), true);
+            if (_newDebt > 0) {
+                addCollateralAndBorrow(_want, _newDebt);
+                uint256 wad = IERC20Upgradeable(debtToken).balanceOf(address(this));
+                _depositUSDP(_newDebt > wad ? wad : _newDebt);
+            }
+        }
+    }
+
+    // to avoid repay all debt resulting to close the CDP unexpectedly
+    function _capMaxDebtPaid(uint256 _actualPaidDebt) internal view returns (uint256) {
+        uint256 _maxDebtToRepay = getDebtWithoutFee().sub(ratioBuffMax);
+        return _actualPaidDebt >= _maxDebtToRepay ? _maxDebtToRepay : _actualPaidDebt;
+    }
+
+    /// @dev Internal logic for partial withdrawals. Should exit positions as efficiently as possible.
+    /// @dev The withdraw() function shell automatically uses idle want in the strategy before attempting to withdraw more using this
+    function _withdrawSome(uint256 _amount) internal override returns (uint256) {
+        if (_amount == 0) {
+            return _amount;
+        }
+
+        uint256 requiredPaidback = requiredPaidDebt(_amount);
+        if (requiredPaidback > 0) {
+            _withdrawUSDP(requiredPaidback);
+        }
+
+        bool _fullWithdraw = _amount == balanceOfPool();
+        uint256 _wantBefore = IERC20Upgradeable(want).balanceOf(address(this));
+        if (!_fullWithdraw) {
+            uint256 _actualPaidDebt = IERC20Upgradeable(debtToken).balanceOf(address(this));
```

Comment:
To avoid wasting gas by calling `IERC20Upgradeable(debtToken).balanceOf(address(this))` again on line 231, which will return the same value, we can store it in a different variable:

```solidity
uint256 _debtTokenBalance = IERC20Upgradeable(debtToken).balanceOf(address(this));
uint256 _fee = getDebtBalance().sub(getDebtWithoutFee());

require(_debtTokenBalance > _fee, "!notEnoughForFee");
uint256 _actualPaidDebt = _debtTokenBalance.sub(_fee); // unit protocol will charge fee first
 _actualPaidDebt = _capMaxDebtPaid(_actualPaidDebt);

require(_debtTokenBalance >= _actualPaidDebt.add(_fee), "!notEnoughRepayment");
```